File: backend/app/hms/pms/schemas/inventory.py
from datetime import date

from pydantic import BaseModel, Field


# ── Availability ─────────────────────────────────────────────────────────────

class AvailabilityRequest(BaseModel):
    check_in: date
    check_out: date
    pax: int = Field(default=1, ge=1, le=20)
    property_id: int = Field(gt=0)


class AvailableRoomRead(BaseModel):
    room_id: int
    room_number: str
    room_type_id: int
    room_type_name: str
    max_occupancy: int
    floor: int | None = None
    status: str


class AvailabilityRead(BaseModel):
    check_in: date
    check_out: date
    nights: int
    pax: int
    rooms: list[AvailableRoomRead] = Field(default_factory=list)


# ── Pricing quote ─────────────────────────────────────────────────────────────

class PricingQuoteRequest(BaseModel):
    check_in: date
    check_out: date
    room_type_id: int = Field(gt=0)
    property_id: int = Field(gt=0)


class RatePlanQuoteRead(BaseModel):
    plan_id: int
    plan_code: str
    plan_name: str
    avg_nightly_rate: float
    total_price: float
    nights: int
    currency: str = "EUR"


class PricingQuoteRead(BaseModel):
    check_in: date
    check_out: date
    nights: int
    room_type_id: int
    rate_plans: list[RatePlanQuoteRead] = Field(default_factory=list)


# ── Hotel Extras ─────────────────────────────────────────────────────────────

class HotelExtraRead(BaseModel):
    id: int
    property_id: int
    name: str
    unit_price: float
    per_person: bool
    daily: bool
    is_active: bool
    sort_order: int


class HotelExtraCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    unit_price: float = Field(ge=0)
    per_person: bool = False
    daily: bool = False
    sort_order: int = 0


class HotelExtraUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    unit_price: float | None = Field(default=None, ge=0)
    per_person: bool | None = None
    daily: bool | None = None
    is_active: bool | None = None
    sort_order: int | None = None


# ── Stay Occupant ─────────────────────────────────────────────────────────────

class StayOccupantRead(BaseModel):
    guest_profile_id: int
    is_primary: bool
    guest_name: str | None = None
    guest_email: str | None = None


class StayOccupantUpsert(BaseModel):
    """Used when saving the Gäste block: replaces all occupants for a stay."""
    occupants: list[int] = Field(
        default_factory=list,
        description="Ordered list of guest_profile_id values. First entry is primary.",
    )

File: frontend/src/features/hms/pms/api/inventory.ts
"use client";

import { getJson, postJson } from "@/lib/api";
import { defaultHotelPropertyId } from "@/lib/hotel-room-types";

// ── Types ─────────────────────────────────────────────────────────────────────

export type AvailableRoom = {
  room_id: number;
  room_number: string;
  room_type_id: number;
  room_type_name: string;
  max_occupancy: number;
  floor: number | null;
  status: string;
};

export type AvailabilityResult = {
  check_in: string;
  check_out: string;
  nights: number;
  pax: number;
  rooms: AvailableRoom[];
};

export type RatePlanQuote = {
  plan_id: number;
  plan_code: string;
  plan_name: string;
  avg_nightly_rate: number;
  total_price: number;
  nights: number;
  currency: string;
};

export type PricingQuoteResult = {
  check_in: string;
  check_out: string;
  nights: number;
  room_type_id: number;
  rate_plans: RatePlanQuote[];
};

export type HotelExtra = {
  id: number;
  property_id: number;
  name: string;
  unit_price: number;
  per_person: boolean;
  daily: boolean;
  is_active: boolean;
  sort_order: number;
};

export type GuestSearchResult = {
  id: number;
  name: string;
  email: string | null;
  phone: string | null;
};

// ── Availability ──────────────────────────────────────────────────────────────

export async function fetchAvailability(
  checkIn: string,
  checkOut: string,
  pax: number,
  propertyId: number = defaultHotelPropertyId,
): Promise<AvailabilityResult> {
  return postJson<AvailabilityResult>("/hms/pms/inventory/availability", {
    check_in: checkIn,
    check_out: checkOut,
    pax,
    property_id: propertyId,
  });
}

// ── Pricing quote ─────────────────────────────────────────────────────────────

export async function fetchPricingQuote(
  checkIn: string,
  checkOut: string,
  roomTypeId: number,
  propertyId: number = defaultHotelPropertyId,
): Promise<PricingQuoteResult> {
  return postJson<PricingQuoteResult>("/hms/pms/pricing/quote", {
    check_in: checkIn,
    check_out: checkOut,
    room_type_id: roomTypeId,
    property_id: propertyId,
  });
}

// ── Extras ────────────────────────────────────────────────────────────────────

export async function fetchExtras(
  propertyId: number = defaultHotelPropertyId,
): Promise<HotelExtra[]> {
  return getJson<HotelExtra[]>("/hms/pms/inventory/extras", {
    params: { property_id: propertyId },
  });
}

// ── Guest search ──────────────────────────────────────────────────────────────

export async function searchGuests(
  query: string,
  propertyId: number = defaultHotelPropertyId,
): Promise<GuestSearchResult[]> {
  if (!query.trim()) return [];
  const contacts = await getJson<
    Array<{ id: number; name: string; email: string | null; phone: string | null }>
  >("/hms/pms/contacts", {
    params: { search: query, property_id: propertyId, limit: 10 },
  });
  return contacts.map((c) => ({
    id: c.id,
    name: c.name,
    email: c.email,
    phone: c.phone,
  }));
}

export async function moveHotelStay(
  stayId: number | string,
  payload: {
    room_id: number;
    notes?: string | null;
  },
  propertyId: number = defaultHotelPropertyId,
) {
  return postJson<{
    reservation_id: number;
    room_id: number | null;
    room_number: string | null;
    room_type_name: string | null;
    stay: {
      id: number;
      property_id: number;
      reservation_id: number;
      room_id: number | null;
      status: string;
      planned_check_in: string;
      planned_check_out: string;
      actual_check_in_at: string | null;
      actual_check_out_at: string | null;
      notes: string | null;
      created_at: string;
      updated_at: string;
    };
  }>(`/hms/stays/${stayId}/move`, payload, {
    params: { property_id: propertyId },
  });
}
